### PR TITLE
fix: add captcha on login and signup page

### LINKED ISSUE #122: Protecting user signup with Captcha
While testing on https://artemis01.sidn.nl/signup I found that there’s no rate limit implemented or CAPTCHA, in this way I can make multiple accounts
Impact:
the attacker can make for example 1 M request that lead to fill your DB with fake accounts .
Step to Reproduce
1. Go to this link and fill up the form
https://artemis01.sidn.nl/signup
or send this request to your Burp Suite Intruder
Code:
POST /api/auth/signup HTTP/1.1
Host: artemis01.sidn.nl
Cookie: sid=1m1mGQmjMEkQIr0KFpUFl
User-Agent: Mozilla/5.0 (Windows NT 6.3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36
Accept: /
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Referer: https://artemis01.sidn.nl/signup
Content-Type: application/json
Origin: https://artemis01.sidn.nl
Content-Length: 169
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-origin
Te: trailers
Connection: close
{“email”:“pusakal+6@wearehackerone.com”,“name”:“gonzx”,“password”:“Gonz@2828",“_csrf”:“GtpTUSd7-ElI51zJ0Z55Cdxc5mTa138TESyc.blNjEilJPmqhR53hYHhdA661HhOc9Ebngy89uX5Dq4U”}
2. Add the email in scope to brute force then start the attack
If you will get a 200 status it means the account is successfully created
Recommendation:
You can use a CAPTCHA

### PR DESCRIPTION
**Pull request by CuriouzK0d3r**
### Description of PR

Resolves #122 

Adds captcha protection to the login and signup pages.

What component(s) does this PR affect?

  - [x] Back-End (Node.js, Next.js)
  - [x] Front-End (React.js)
  - [ ] Theming (CSS, React theme)
  - [ ] Build System
  - [x] Docs
  
  

  Screenshot 2021-09-21 at 1 35 52 PM
Screenshot 2021-09-21 at 1 35 40 PM

  Does the PR require changes on other components? If yes, please mark the components:
  
  - [ ] Back-End (Node.js, Next.js)
  - [ ] Front-End (React.js)
  - [ ] Theming (CSS, React theme)
  - [ ] Build System
  - [ ] Docs
  
  
  
### Related Issue
  
  
  
  
  
  
  Resolves #
  
### Type
  
  - [ ] Bug fix (non-breaking change which fixes an issue)
  - [ ] New feature (non-breaking change which adds functionality)
  - [ ] Breaking change (fix or feature that would cause existing functionality to change)
  - [ ] Docs update
  - [ ] None of the above
  
  
  
### Checklist:
  
  - [ ] I have read the **[contributing guide](https://github.com/FORTH-ICS-INSPIRE/artemis/blob/master/CONTRIBUTING.md)** and my code conforms to the guidelines.
  - [ ] This change requires a change in the documentation.
  - [ ] I have updated the documentation accordingly.
  >)`>>>)

>

### CODE REVIEW

File: apps/artemis-web/pages/api/auth/login/ldap.ts [hunk 1]
```diff
@@ -7,14 +7,22 @@ import {
...
+  .use(captcha('login'))
   .use(auth)
+  .post(async (req: NextApiRequestExtended, res: NextApiResponseExtended, next) => {
+    memory.incr(req.ip);
+    next();
+  })
   .post(
     passport.authenticate('ldapauth'),
     (req: NextApiRequestExtended, res: NextApiResponseExtended, next) => {
+      memory.decr(req.ip);
```
Comment (vkotronis) [inline apps/artemis-web/pages/api/auth/login/ldap.ts:25]: post-login, shouldn't the captcha failed failed try counter be set to 0? Since from now on the user succeeded?

Comment (CuriouzK0d3r) [reply apps/artemis-web/pages/api/auth/login/ldap.ts:25]: fair point! Let me fix that!

File: apps/artemis-web/middleware/captcha.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+const captchaChallengeStore = {};
+
+export default function captcha(page: string) {
+    return async (req, res, next): Promise<any> => {
+        const captchaResult = lambdaCaptcha.verify(req.body.encryptedExpr, req.body.captcha, SECRET);
+        const hits = memory.getHits(req.ip);
+        if (!captchaChallengeStore[req.ip])
+            captchaChallengeStore[req.ip] = [];
+        const skipCaptcha = (page === 'login' && hits < (parseInt(process.env.CAPTCHA_TRIES ?? '4'))) || (process.env.TESTING === 'true');
```
Comment (vkotronis) [inline apps/artemis-web/middleware/captcha.ts:14]: what if hits is undefined? Can it be or is it initialized to 0 by default?

File: apps/artemis-web/libs/passport-strategies/local-strategy.ts [hunk 3]
```diff
@@ -24,6 +25,7 @@ export const LocalStrategy = new Strategy(
         );
```
Comment (vkotronis) [inline apps/artemis-web/libs/passport-strategies/local-strategy.ts:25]: hmm here since the user is authenticated (local strategy), shouldn't also the counter be reset?

File: apps/artemis-web/utils/captchaMemoryStore.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        if (hits[key] && hits[key] !== 0) {
+            hits[key]--;
+        }
+
+        return hits[key];
+    };
+
+    this.incr = function (key) {
+        if (hits[key]) {
+            hits[key]++;
```
Comment (vkotronis) [inline apps/artemis-web/utils/captchaMemoryStore.ts:14]: this should happen only if `hits[key] < CAPTCHA_TRIES`. Otherwise we may increase the counter to inifinity, which we do not need to. If `hits` reached the max limit, then it can stay there; the captcha will be activated until the counter is reset or decreased anyway (reset seems a better solution for captcha since it is an on-off feature; if you login OK, then reset either via LDAP or locally, otherwise increase until limit is hit)

File: apps/artemis-web/middleware/captcha.ts (hunk 2 above)
Comment (CuriouzK0d3r) [reply apps/artemis-web/middleware/captcha.ts:14]: I think you are right. Added a fallback

File: apps/artemis-web/utils/captchaMemoryStore.ts (hunk 4 above)
Comment (CuriouzK0d3r) [reply apps/artemis-web/utils/captchaMemoryStore.ts]: fixed :)

File: apps/artemis-web/libs/passport-strategies/local-strategy.ts (hunk 3 above)
Comment (CuriouzK0d3r) [reply apps/artemis-web/libs/passport-strategies/local-strategy.ts:25]: I added this reset to the /credentials.ts and /ldap.ts files

File: apps/artemis-web/components/hijack-info/hijack-info.tsx [hunk 5]
```diff
@@ -104,7 +103,208 @@ class HijackInfoComponent extends Component<any, any> {
...
+              </div>
+            </div>
+          </div>
+        </div>
+
+        <div className="row">
+          <div className="col-lg-12">
+            <div className="card" style={{ marginTop: '12px', visibility: this.state.gripState ? "visible" : "hidden" }}>
+              <div className="card-header">
+                CAIDA GRIP{' '}
```
Comment (vkotronis) [inline apps/artemis-web/components/hijack-info/hijack-info.tsx:268]: could you rename to `Also detected by the GRIP project` (minor)? Instead of CAIDA GRIP.

File: apps/artemis-web/components/hijack-info/hijack-info.tsx [hunk 6]
```diff
@@ -104,7 +103,208 @@ class HijackInfoComponent extends Component<any, any> {
       return "all";
   }
 
-  render() {
+  getMobileInfo(mRef, hijackKey, commentRef, event_data, type): any {
```
Comment (vkotronis) [inline apps/artemis-web/components/hijack-info/hijack-info.tsx:106]: why this large change here?

File: apps/artemis-web/components/hijack-info/hijack-info.tsx [hunk 7]
```diff
@@ -116,7 +316,7 @@ class HijackInfoComponent extends Component<any, any> {
 
     const type = this.getEventType(this.props.hijackDataState["type"]);
     const asn = this.props.hijackDataState["hijack_as"];
-    const prefix = this.props.hijackDataState["prefix"];
+    // const prefix = this.props.hijackDataState["prefix"];
```
Comment (vkotronis) [inline apps/artemis-web/components/hijack-info/hijack-info.tsx:319]: why is this commented out?

File: apps/artemis-web/components/hijack-info/hijack-info.tsx (hunk 6 above)
Comment (CuriouzK0d3r) [reply apps/artemis-web/components/hijack-info/hijack-info.tsx:106]: Because I was trying to pass the code smells :P

File: apps/artemis-web/components/hijack-info/hijack-info.tsx (hunk 7 above)
Comment (CuriouzK0d3r) [reply apps/artemis-web/components/hijack-info/hijack-info.tsx:319]: It was not being used

File: apps/artemis-web/components/hijack-info/hijack-info.tsx (hunk 5 above)
Comment (vkotronis) [reply apps/artemis-web/components/hijack-info/hijack-info.tsx:268]: Also please do one more renaming because somehow it slipped through the cracks: in the hijack info page, it should be "Hijacker AS" not "Hijacked AS" (important)

File: apps/artemis-web/components/hijack-info/hijack-info.tsx (hunk 6 above)
Comment (vkotronis) [reply apps/artemis-web/components/hijack-info/hijack-info.tsx:106]: does it change the logic, or is it simply a formatting change? hard to understand with a large PR :)

File: apps/artemis-web/components/hijack-info/hijack-info.tsx (hunk 7 above)
Comment (vkotronis) [reply apps/artemis-web/components/hijack-info/hijack-info.tsx:319]: and why is only the `asn` and `type` used? To check for GRIP fetching we need all type, asn and prefix to be actually present.

File: apps/artemis-web/libs/passport-strategies/local-strategy.ts [hunk 8]
```diff
@@ -23,6 +23,7 @@ export const LocalStrategy = new Strategy(
             },
           }
         );
+        memory.reset(req.ip);
```
Comment (vkotronis) [inline apps/artemis-web/libs/passport-strategies/local-strategy.ts:26]: so why is this needed here and in the credentials? Is it because credentials are not actually visited before someone presses the login button?

File: apps/artemis-web/components/hijack-info/hijack-info.tsx (hunk 6 above)
Comment (CuriouzK0d3r) [reply apps/artemis-web/components/hijack-info/hijack-info.tsx:106]: no it is just method extraction. I moved the HTML to a new method

File: apps/artemis-web/components/hijack-info/hijack-info.tsx (hunk 7 above)
Comment (CuriouzK0d3r) [reply apps/artemis-web/components/hijack-info/hijack-info.tsx:319]: You are very right. However, this logic is included in   async fetchGrip(hijackDataState) {
 function.

The data are stored in hijackDataState. Prefix var was redundant

Comment (vkotronis) [reply apps/artemis-web/components/hijack-info/hijack-info.tsx:319]: I see, now it is clear. Resolving this pending comment.

**Review by vkotronis — APPROVED** [review]
tested, LGTM! Good work @CuriouzK0d3r !
